### PR TITLE
choleskify vs triangularize

### LINKED ISSUE #20: Should `choleskify` return `Cholesky` type?
Currently, `choleskify_factor` turns any PSD matrix into a PSDMatrix, where the factor `R` is a valid Cholesky factor. 

I personally think `choleskify_factor` should be split into a `triangularize_factor()` method, which turns any `PSDMatrix` into one with an `UpperTriangular` factor, and a `choleskify` method, which turns a `PSDMatrix` with an `UpperTriangular` factor into a `Cholesky` type. Why?

* A method that is called `choleskify` would be easier to understand if it returned a Cholesky type
* In most of our applications, triangular factors are sufficient, and the extra multiplication that turns upper triangular types into valid Cholesky factors can be avoided

Thoughts, @nathanaelbosch?

### PR DESCRIPTION
**Pull request by pnkraemer**
If you agree with #20 (splitting `choleskify_factors` into `triangularize_factors` and `choleskify`), here is an implementation. Personally, I prefer it this way. But if you disagree, no worries, then I will just delete this.

### CODE REVIEW

File: src/PSDMatrices.jl [hunk 1]
```diff
@@ -78,19 +72,33 @@ end
...
     R = qr(M.R).R
-    R = nonnegative_diagonal(R)
     return PSDMatrix(UpperTriangular(R))
 end
 
+function choleskify(M::PSDMatrix)
+    if M.R isa UpperTriangular
+        return Cholesky(nonnegative_diagonal(M.R), 'U', 0)
+    end
+    errormsg = "choleskify() is only defined for PSDMatrix types with upper triangular factors."
```
Comment (nathanaelbosch) [inline src/PSDMatrices.jl]: Why should this raise an error? I think I'd generally be in favor of providing a `choleskify` (or `Cholesky`, as commented in #20) not just for `PSDMatrix`es with triangular factors, but which also triangularize the factors if necessary.

File: test/runtests.jl [hunk 2]
```diff
@@ -63,8 +63,16 @@ eltypes = (Int64, Float64, BigFloat)
...
             if (size(M, 1) >= size(M, 2))
-                @test choleskify_factor(S).R ≈ cholesky(Matrix(S)).U
                 @test Matrix(add_cholesky(S, S)) ≈ Matrix(S) + Matrix(S)
+
+                tri = triangularize_factor(S)
+                chol = cholesky(Matrix(S))
+                # Assert that `tri` is basically a Cholesky factorisation, 
+                # but without guaranteed positive diagonal elements
+                @test begin
+                    tri.R isa UpperTriangular
```
Comment (nathanaelbosch) [inline test/runtests.jl]: How about simplifying this to just
```
tri = triangularize_factor(S)
@test tri.R*tri.R' \approx Matrix(S)
```
? Or do we need more?

File: test/runtests.jl [hunk 3]
```diff
@@ -63,8 +63,16 @@ eltypes = (Int64, Float64, BigFloat)
...
+                    tri.R' * tri.R ≈ chol.U' * chol.U
+                end
```
Comment (pnkraemer) [inline test/runtests.jl]: This should actually suffice
```suggestion
                tri = triangularize_factor(S)
                @test tri.R isa UpperTriangular
```

Comment (nathanaelbosch) [reply test/runtests.jl]: But this does not actually test if the matrix is still correct? E.g. a `triangularize_factor(S) = PSDMatrix(UpperTriangular(rand(2,2)))` would pass the test?

File: test/runtests.jl [hunk 4]
```diff
@@ -63,8 +63,9 @@ eltypes = (Int64, Float64, BigFloat)
             end
             @test Matrix(add_qr(S, S)) ≈ Matrix(S) + Matrix(S)
             if (size(M, 1) >= size(M, 2))
-                @test choleskify_factor(S).R ≈ cholesky(Matrix(S)).U
                 @test Matrix(add_cholesky(S, S)) ≈ Matrix(S) + Matrix(S)
+                tri = triangularize_factor(S)
+                @test tri.R isa UpperTriangular
```
Comment (pnkraemer) [inline test/runtests.jl:68]:
```suggestion
                @test tri.R isa UpperTriangular
                @test Matrix(tri) ≈ Matrix(S)
```

**Review by nathanaelbosch — APPROVED** [review]

**Pull request merged**